src/sentry_core.c

```diff
@@ -772,20 +773,26 @@ sentry_transaction_start(sentry_value_t tx_cxt)
 sentry_uuid_t
 sentry_transaction_finish(sentry_value_t tx)
 {
-    if (sentry_value_get_type(tx) == SENTRY_VALUE_TYPE_STRING) {
-        sentry_value_decref(tx);
-        tx = sentry_value_new_null();
-        // Minimize the amount of time spent locking the scope, and do checking
-        // on the transaction when this is not holding onto the lock.
-        SENTRY_WITH_SCOPE (scope) {
-            tx = sentry__value_clone(scope->span);
-        }
-        sentry__scope_remove_span();
-    }
     if (sentry_value_is_null(tx)) {
         SENTRY_DEBUG("no transaction available to finish");
         goto fail;
     }
+
+    SENTRY_WITH_SCOPE_MUT_NO_FLUSH (scope) {
+        const char *tx_id
+            = sentry_value_as_string(sentry_value_get_by_key(tx, "trace_id"));
+        const char *scope_tx_id = sentry_value_as_string(
+            sentry_value_get_by_key(scope->span, "trace_id"));
+        if (sentry__string_eq(tx_id, scope_tx_id)) {
+            sentry_value_decref(scope->span);
+            scope->span = sentry_value_new_null();
+        }
+    }
+    // immediately flush the scope so the transaction doesn't get attached to
+    // any new events
+    SENTRY_WITH_SCOPE_MUT (scope) {
```

Comment:
You have this following immediately a `SENTRY_WITH_SCOPE_MUT_NO_FLUSH`. Seems kind of backwards to me? Whats the reason for not doing a `MUT` right away?